### ISSUE TITLE
Throttling issue

### ISSUE THREAD

**Issue by jelleholtkamp** [microsoftgraph/msgraph-sdk-powershell#1085]
I'm working on a script that checks if a given SPN is still being used to sign in. First I grab all SPN's with the Get-AzureADServicePrincipal -All $true powershell cmdlet, then I loop through them and call the following URI:

https://graph.microsoft.com/beta/auditLogs/signIns?`$filter=(servicePrincipalId eq '$spnId' and signInEventTypes/any(t: t eq 'servicePrincipal'))

After about 7 loops, I'm already getting a 429 Too Many Requests error. I added a sleep of 2,5 seconds to each loop and that helps (so the issue seems to be too many requests in x seconds/milliseconds and not hours), though it will still give me a 429 after about 400 loops. I'm currently seeing if I can make it through the whole thing if I add a 3 second sleep.

I'm using a similar call to https://graph.microsoft.com/beta/users/$UPN/authentication/methods to get MFA details for our users, and I'm not getting any throttling issues there. It seems to me that the /auditLogs endpoint has very stricts throttling policies compared to other endpoints. Is there a reason to this? Will this be changed once this endpoint gets out of beta?

**Comment by insomniacc** [comment]
@jelleholtkamp  - Havent used this personally yet, but I believe the response header should give a retry time? If so you can use this in your error handling to back off for a set time. I may be wrong but I also believe retries to the API - even failed ones will also count towards throttling.
I implement an incremental back off in my calls, basically I wait 3sec, if I get another failure, I wait 6sec and then 9 sec until eventually I'll throw if it still doesnt get through, seems to work fine for me for now.

**Comment by jelleholtkamp** [comment]
@insomniacc That's what I ended up doing as well, as well as a 2 second delay every loop. That seems to work for now, however I still think it's odd that just this specific call/endpoint has such a tight throttling policy, while with other calls I don't run into any throttling isseus.

**Comment by peombwa** [comment]
@jelleholtkamp, is there a reason you are not using the [Microsoft Graph PowerShell SDK](https://docs.microsoft.com/en-us/graph/powershell/installation?view=graph-rest-1.0)? The SDK handles throttling for you out of the box as described in https://microsoftgraph.github.io/msgraph-sdk-design/middleware/RetryHandler.html.
``` powershell
Select-MgProfile beta
Get-MgAuditLogSignIn -Filter "(servicePrincipalId eq '$spnId' and signInEventTypes/any(t: t eq 'servicePrincipal'))"
```

With Microsoft Graph, each service sets its throttling limit as described here at https://docs.microsoft.com/en-us/graph/throttling?context=graph%2Fapi%2F1.0&view=graph-rest-1.0#service-specific-limits.

**Comment by jelleholtkamp** [comment]
@peombwa I didn't find out about the SDK untill I started looking into this issue. I'll try the SDK to see if I can improve the handling of the throttling in a better manner. 

The documents you provided have been helpful. I was unsure if the throttling limits I encountered were intentional, but it seems they are :

<html>
<body>
<h3 id="identity-and-access-reports-service-limits" class="heading-anchor" style="box-sizing: inherit; margin: 30px 0px 18px -1.875rem; padding: 0px 0px 0px 1.875rem; font-size: clamp(1.3125rem, 15.4737px + 1.15132vw, 1.75rem); font-weight: 600; outline-color: inherit; line-height: 1.3; position: relative; color: rgb(23, 23, 23); font-family: &quot;Segoe UI&quot;, SegoeUI, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Identity and access reports service limits</h3><div class="table-scroll-wrapper has-inner-focus" style="box-sizing: inherit; outline-color: inherit; overflow: auto hidden; color: rgb(23, 23, 23); font-family: &quot;Segoe UI&quot;, SegoeUI, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Request type | Limit per app per tenant
-- | --
Any | 5 requests per 10 seconds

</div>
</body>
</html>

Those limits are consistent with what I've been experiencing. Thanks for the input.
